// @expo/cli/build/src/start/server/AsyncWsTunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AsyncWsTunnel", {
    enumerable: true,
    get: ()=>AsyncWsTunnel
});
function _wsTunnel() {
    const data = /*#__PURE__*/ _interopRequireWildcard(require("@expo/ws-tunnel"));
    _wsTunnel = function() {
        return data;
    };
    return data;
}
function _chalk() {
    const data = /*#__PURE__*/ _interopRequireDefault(require("chalk"));
    _chalk = function() {
        return data;
    };
    return data;
}
function _nodeCrypto() {
    const data = require("node:crypto");
    _nodeCrypto = function() {
        return data;
    };
    return data;
}
const _log = /*#__PURE__*/ _interopRequireWildcard(require("../../log"));
const _env = require("../../utils/env");
const _errors = require("../../utils/errors");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const debug = require("debug")("expo:start:server:ws-tunnel");
class AsyncWsTunnel {
    /** Info about the currently running instance of tunnel. */ serverUrl = null;
    constructor(_projectRoot, port){
        if (port !== 8081) {
            throw new _errors.CommandError("WS_TUNNEL_PORT", `WS-tunnel only supports tunneling over port 8081, attempted to use port ${port}`);
        }
    }
    getActiveUrl() {
        return this.serverUrl;
    }
    async startAsync() {
        this.serverUrl = await _wsTunnel().startAsync({
            ...getTunnelOptions(),
            onStatusChange (status) {
                if (status === "disconnected") {
                    _log.error(_chalk().default.red("Tunnel connection has been closed. This is often related to intermittent connection problems with the ws proxy servers. Restart the dev server to try connecting again.") + _chalk().default.gray("\nCheck the Expo status page for outages: https://status.expo.dev/"));
                }
            }
        });
        debug("Tunnel URL:", this.serverUrl);
    }
    async stopAsync() {
        debug("Stopping Tunnel");
        await _wsTunnel().stopAsync();
        this.serverUrl = null;
    }
}
function getTunnelOptions() {
    const userDefinedSubdomain = _env.env.EXPO_TUNNEL_SUBDOMAIN;
    if (userDefinedSubdomain && typeof userDefinedSubdomain === "string") {
        debug("Session:", userDefinedSubdomain);
        return {
            session: userDefinedSubdomain
        };
    }
    let session;
    do {
        // TODO(cedric): replace this with non-random data generated from server to manage and prevent overlapping sessions
        session = (0, _nodeCrypto().randomBytes)(12).toString("base64url");
    }while (!/^[A-Za-z0-9]/.test(session));
    debug("Session:", session);
    return {
        session
    };
}
